drivers/sensor/bh1749/bh1749.c

```diff
@@ -0,0 +1,287 @@
+		break;
+	case SENSOR_CHAN_BLUE:
+		val->val1 = sys_le16_to_cpu(
+			data->sample_rgb_ir[BH1749_RGB_BYTE_POS_BLUE]);
+		val->val2 = 0;
+		break;
+	case SENSOR_CHAN_IR:
+		val->val1 = sys_le16_to_cpu(
+			data->sample_rgb_ir[BH1749_RGB_BYTE_POS_IR]);
+		val->val2 = 0;
+		break;
+	default:
+		return -ENOTSUP;
+	}
+
+	return 0;
+}
+
+static int bh1749_check(struct device *dev)
+{
+	int err;
+	struct bh1749_data *data = dev->driver_data;
+
+	u8_t manufacturer_id;
+
+	err = i2c_reg_read_byte(data->i2c, DT_ROHM_BH1749_0_BASE_ADDRESS,
+				BH1749_MANUFACTURER_ID, &manufacturer_id);
+	if (err) {
+		LOG_ERR("Failed when reading manufacturer ID");
+		return -EIO;
+	}
+
+	LOG_DBG("Manufacturer ID: 0x%02x", manufacturer_id);
+
+	if (manufacturer_id != BH1749_MANUFACTURER_ID_DEFAULT) {
+		LOG_ERR("Invalid manufacturer ID: 0x%02x", manufacturer_id);
+		return -EIO;
+	}
+
+	u8_t part_id;
+
+	err = i2c_reg_read_byte(data->i2c, DT_ROHM_BH1749_0_BASE_ADDRESS,
+				BH1749_SYSTEM_CONTROL, &part_id);
+	if (err) {
+		LOG_ERR("Failed when reading part ID");
+		return -EIO;
+	}
+
+	if ((part_id & BH1749_SYSTEM_CONTROL_PART_ID_Msk) !=
+	    BH1749_SYSTEM_CONTROL_PART_ID) {
+		LOG_ERR("Invalid part ID: 0x%02x", part_id);
+		return -EIO;
+	}
+
+	LOG_DBG("Part ID: 0x%02x", part_id);
+
+	return 0;
+}
+
+static int bh1749_sw_reset(struct device *dev)
+{
+	return i2c_reg_update_byte(dev, DT_ROHM_BH1749_0_BASE_ADDRESS,
+				   BH1749_SYSTEM_CONTROL,
+				   BH1749_SYSTEM_CONTROL_SW_RESET_Msk,
+				   BH1749_SYSTEM_CONTROL_SW_RESET);
+}
+
+static int bh1749_rgb_gain_set(struct device *dev, u8_t value)
+{
+	return i2c_reg_update_byte(dev, DT_ROHM_BH1749_0_BASE_ADDRESS,
+				   BH1749_MODE_CONTROL1,
+				   BH1749_MODE_CONTROL1_RGB_GAIN_Msk,
+				   value);
+}
+
+static int bh1749_rgb_measurement_enable(struct device *dev, bool enable)
+{
+	u8_t en = enable ?
+		  BH1749_MODE_CONTROL2_RGB_EN_ENABLE :
+		  BH1749_MODE_CONTROL2_RGB_EN_DISABLE;
+
+	return i2c_reg_update_byte(dev, DT_ROHM_BH1749_0_BASE_ADDRESS,
+				   BH1749_MODE_CONTROL2,
+				   BH1749_MODE_CONTROL2_RGB_EN_Msk,
+				   en);
+}
+
+static int bh1749_init(struct device *dev)
+{
+	int err;
+	struct bh1749_data *data = dev->driver_data;
+
+	/* Activation time 2ms */
+	k_sleep(2);
+
+	data->i2c = device_get_binding(DT_ROHM_BH1749_0_BUS_NAME);
+
+	if (data->i2c == NULL) {
+		LOG_ERR("Failed to get pointer to %s device!",
+			DT_ROHM_BH1749_0_BUS_NAME);
+		return -EINVAL;
+	}
+
+	(void)memset(data->sample_rgb_ir, 0, sizeof(data->sample_rgb_ir));
+	data->pdata = 0U;
+
+	err = bh1749_sw_reset(data->i2c);
+	if (err) {
+		LOG_ERR("Could not apply software reset.");
+		return -EIO;
+	}
+
+	err = bh1749_check(dev);
+	if (err) {
+		LOG_ERR("Communication with BH1749 failed with error %d", err);
+		return -EIO;
+	}
+
+	err = bh1749_rgb_measurement_enable(data->i2c, true);
+	if (err) {
+		LOG_ERR("Could not set measurement mode.");
+		return -EIO;
+	}
+
+
+	err = bh1749_rgb_gain_set(data->i2c, BH1749_MODE_CONTROL1_RGB_GAIN_1X);
```

Comment:
Wouldn't it be better to set initial values of all the fields of this register in one call? The default value of this register `0`, so after setting only the RGB gain and the measurement mode, the IR gain will stay at `0`, which is forbidden to use.

Reply:
I can do that. I'll remove the rgb_gain_set function then

Reply:
Replaced mode and gain set with
`	if (i2c_reg_write_byte(data->i2c, DT_ROHM_BH1749_0_BASE_ADDRESS,
			       BH1749_MODE_CONTROL1,
			       BH1749_MODE_CONTROL1_DEFAULTS)) {
		LOG_ERR("Could not set gain and measurement mode configuration.");
		return -EIO;
	}`